### PR TITLE
Faster REPL - don't load extra middleware on first sync eval request

### PR DESCRIPTION
**Pull request by vspinu**
This saves a couple of extra seconds and brings cider repl startup time to barebones empty-project `lein repl` of 4s for me.

----

  - Send :inhibit-cider-middleware condition on first request. Relies on a feature added in clojure-emacs/cider-nrepl#438.

  - Combine `require-repl-utils` and `set-initial-ns` into one sync request for efficiency. Note that `require-repl-utils` needs to be a sync request. It could happen that the following init-debuger request would occur before it has finished, resulting in "reader/reader not found error".

### CODE REVIEW

**Comment by vspinu** [comment]
From #2081:
> I was just wondering whether to try to include there some of @vspinu's performance improvement patches or push those back to 0.16.

It depends how long would it take to release 0.16.  I would like to see #2069 in 0.16 and that change with associated features/fixes might take quite some months. If we can pull it in within a couple of months then this change can also go in 0.16, if not then it might be better included now.

As noted in 721440b99a23947, there is one side effect of dynamic loading of middlware - it can occur concurrently. But Clojure seems to have trouble loading namespaces in parallel.

I haven't investigated this thoroughly but I run into issues when debug middleware was requested before `out` middleware on emacs side in async requests. As debug is heavy, `out` middleware ends up loaded simultaneously with `debug` and didn't load correctly.   This is why I changed the order of those two requests on emacs side. `out` middleware is very light and doesn't cause issues with the debug. 

Fixing this on cider's side is easy, just implement a kind of deffered chain to load cider's middleware. But then, some other packages (nrepl-refactor?) might load same namespaces in parallel. One partial solution is to run `cider-connected-hook` at the end of the deffered chain and request external packages to rely on `cider-connected-hook` (which they should do anyways).

**Comment by xiongtx** [comment]
> But Clojure seems to have trouble loading namespaces in parallel.

**Comment by vspinu** [comment]
> Is this due to dependencies between namespaces? Needs a topological sort?

No. It's just that if two actions simultaneously try to load same namespace, something goes wrong and one of them is not loaded correctly.

**Comment by xiongtx** [comment]
Can you put up a minimal repo that can reproduce this problem? Seems like a Clojure level issue, not `cider-nrepl`.

**Comment by vspinu** [comment]
Yes, it's a Clojure level issue. If I had a minimal rep. example I would have already reported it on JIRA.

**Comment by bbatsov** [comment]
> As noted in 721440b, there is one side effect of dynamic loading of middlware - it can occur concurrently. But Clojure seems to have trouble loading namespaces in parallel.

Things are these are generally my concern - don't like introducing potentially breaking changes in a bugfix release. 

> It depends how long would it take to release 0.16. I would like to see #2069 in 0.16 and that change with associated features/fixes might take quite some months. If we can pull it in within a couple of months then this change can also go in 0.16, if not then it might be better included now.

There's never a precise roadmap for releases. Generally I cut them when I feel we've added something important enough to warrant a release.

File: cider-repl.el [hunk 1]
```diff
@@ -255,26 +255,24 @@ ENDPOINT is a plist as returned by `nrepl-connect'."
...
-   "(when (clojure.core/resolve 'clojure.main/repl-requires)
-      (clojure.core/map clojure.core/require clojure.main/repl-requires))"
-   (lambda (_response) nil)))
-
 (declare-function cider-set-buffer-ns "cider-mode")
-(defun cider-repl-set-initial-ns (buffer)
-  "Set the REPL BUFFER's initial namespace (by altering `cider-buffer-ns').
-This is \"user\" by default but can be overridden in apps like lein (:init-ns)."
+(defun cider-repl-require-repl-utils-and-set-ns (buffer)
+  "Require standard REPL util functions and set the REPL BUFFER NS.
```
Comment (bbatsov) [inline cider-repl.el]: NS -> ns (it's not a param).

File: cider-repl.el [hunk 2]
```diff
@@ -255,26 +255,24 @@ ENDPOINT is a plist as returned by `nrepl-connect'."
...
+Namespace is \"user\" by default but can be overridden in apps like
```
Comment (bbatsov) [inline cider-repl.el]: default, but

File: cider-repl.el [hunk 3]
```diff
@@ -255,26 +255,24 @@ ENDPOINT is a plist as returned by `nrepl-connect'."
...
+lein (:init-ns). Booth of these operations need to be done as a sync
```
Comment (bbatsov) [inline cider-repl.el]: booth -> both

File: cider-repl.el [hunk 4]
```diff
@@ -255,26 +255,24 @@ ENDPOINT is a plist as returned by `nrepl-connect'."
...
+request at the beginning of the session. Bundling them for efficiency."
```
Comment (bbatsov) [inline cider-repl.el]: There's one space missing here before the final sentence. 
them -> them together

File: cider-repl.el [hunk 5]
```diff
@@ -289,18 +287,29 @@ to call `cider-remember-clojure-buffer'.")
...
-  (cider-repl-set-initial-ns buffer)
-  (cider-repl-require-repl-utils)
-  (unless no-banner
-    (cider-repl--insert-banner-and-prompt buffer))
   (when cider-repl-display-in-current-window
     (add-to-list 'same-window-buffer-names (buffer-name buffer)))
   (pcase cider-repl-pop-to-buffer-on-connect
     (`display-only (display-buffer buffer))
     ((pred identity) (pop-to-buffer buffer)))
+  (cider-repl-require-repl-utils-and-set-ns buffer)
```
Comment (bbatsov) [inline cider-repl.el]: Why did you move those down?

File: cider.el [hunk 6]
```diff
@@ -892,10 +892,15 @@ buffer."
     (cider--check-required-nrepl-version)
     (cider--check-clojure-version-supported)
     (cider--check-middleware-compatibility)
-    (cider--debug-init-connection)
     (cider--subscribe-repl-to-server-out)
     (when cider-auto-mode
       (cider-enable-on-existing-clojure-buffers))
+    ;; Middleware on cider-nrepl side is loaded dynamically on first
+    ;; usage. Loading middleware concurrently can lead to confusing issues
+    ;; (likely a clojure bug). Debug middleware is heavy, so enable it at the
```
Comment (bbatsov) [inline cider.el]: I recall you verified this is a bug, so I think you should mention it here. The next sentence doesn't sound very reassuring. 😃

**Comment by bbatsov** [comment]
I guess you can also add to this PR a changelog entry, as users are going to be quite excited about the speedup.

**Comment by bbatsov** [comment]
@vspinu ping :-)

**Comment by vspinu** [comment]
Sorry for being slow on this. My kid was just born a couple of weeks ago, so my bandwidth is 1/10th of what it was before :) Will try to fix all my PRs by Saturday.

**Comment by bbatsov** [comment]
> Sorry for being slow on this. My kid was just born a couple of weeks ago, so my bandwidth is 1/10th of what it was before :) Will try to fix all my PRs by Saturday.

Thanks and congratulations! 🎉 🎂

File: cider-repl.el (hunk 5 above)
Comment (vspinu) [reply cider-repl.el]: The REPL shows a bit earlier (.5-1 sec for me). Also, the repl is always shown, even when errors occur with those subsequent requests.

**Comment by vspinu** [comment]
@bbatsov, I have addressed your comments in all of my PRs (also the one on cider-nrepl side). Added all change-log entries here to avoid conflicts.

**Comment by bbatsov** [comment]
👍
